### PR TITLE
core/tracker: refactor inclusion checker

### PR DESCRIPTION
**Pull request by corverroos**
Refactors the inclusion delay tracker to also detect missed inclusions and to log broadcast delay of included and missed duties. The aim is to find a correlation between broadcast performance and inclusion delay.

category: feature
ticket: #1538

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -398,9 +398,13 @@ func wireCoreWorkflow(ctx context.Context, life *lifecycle.Manager, conf Config,
 	feeRecipientFunc := func(pubkey core.PubKey) string {
 		return feeRecipientAddrByCorePubkey[pubkey]
 	}
-
 	sched.SubscribeSlots(setFeeRecipient(eth2Cl, eth2Pubkeys, feeRecipientFunc))
-	sched.SubscribeSlots(tracker.NewInclDelayFunc(eth2Cl, sched.GetDutyDefinition))
+
+	inclDelay, err := tracker.NewInclusionDelay(ctx, eth2Cl, sched.GetDutyDefinition)
+	if err != nil {
+		return err
```
Comment (dB2510) [inline app/app.go:405]: nit: can you wrap this error just for the sake of context?

File: core/tracker/incldelay.go [hunk 2]
```diff
@@ -6,112 +6,211 @@ import (
...
+	dutiesFunc     dutiesFunc
+	instrumentFunc func(inclDelaySlots []int64)
+	logMappingFunc func(ctx context.Context, slot int64, bcastDelay time.Duration, inclDelay int64, isAggregated bool)
+	genesisTime    time.Time
+	slotDuration   time.Duration
+	slotsPerEpoch  int
+	clock          clockwork.Clock
+
+	mu            sync.Mutex
+	bcastDelays   map[int64]map[eth2p0.Root]time.Duration
```
Comment (dB2510) [inline core/tracker/incldelay.go:75]: so we are storing broadcast delays for attestations per slot per block root, but we can submit multiple attestations with the same slot and block root. 

This map will only take care of one of those attestation. I think if we are tracking then we should track all of them. Maybe modify the map as `map[int64]map[eth2p0.BLSPubKey]time.Duration` where we can get the pubkey from duty.

Comment (dB2510) [reply core/tracker/incldelay.go:75]: Also while iterating all the attestation in the block this broadcast might not EXACTLY map with the desired attestation if we have submitted multiple attestations with the same and block root.

File: core/tracker/inclusion.go [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+		}
+
+		switch sub.Duty.Type {
+		case core.DutyAttester, core.DutyAggregator:
+			_, ok := block.Attestations[sub.AttRoot]
+			if !ok {
+				continue
+			}
+			i.submissions[j].Included = true
+			i.attIncludedFunc(ctx, sub, block)
```
Comment (dB2510) [inline core/tracker/inclusion.go:144]: we need to also check the case where our single unaggregated attestation is included as part of aggregated attestation. Here you are checking if our unaggregated attestation is included or not OR aggregated attestation submitted by us is included or not.

AttRoot of an unaggregated attestation will have aggregation bits with only one bit set. While we need to check for the aggregation bits of an aggregated block attestation at the validator committee index like we were doing earlier.

File: core/tracker/inclusion.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+			z.Any("pubkey", sub.Pubkey),
+			z.I64("block_slot", sub.Duty.Slot),
+			z.Any("broadcast_delay", sub.Delay),
+		)
+	default:
+		panic("bug: unexpected type") // Sanity check, this should never happen
+	}
+}
+
+func reportAttInclusion(ctx context.Context, sub submission, block block) {
```
Comment (dB2510) [inline core/tracker/inclusion.go:189]: no need to pass full block, just pass `blockSlot`. `AttSlot` can be taken from submission.

Comment (dB2510) [reply core/tracker/inclusion.go:189]: Also add godoc

File: core/tracker/inclusion.go [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+		}
+		if !sub.Included {
+			i.missedFunc(ctx, sub)
+		}
+	}
+	i.submissions = remaining
+}
+
+// CheckBlock checks whether the block includes any of the submitted duties.
+func (i *inclusion) CheckBlock(ctx context.Context, block block) {
```
Comment (dB2510) [inline core/tracker/inclusion.go:128]: why this is a public method?

File: core/tracker/inclusion.go [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+			missedFunc:      reportMissed,
+		},
+		eth2Cl:       eth2Cl,
+		genesis:      genesis,
+		slotDuration: slotDuration,
+	}, nil
+}
+
+// InclusionChecker checks whether duties have been included in blocks.
+type InclusionChecker struct {
```
Comment (dB2510) [inline core/tracker/inclusion.go:234]: why you needed to wrap inclusion with InclusionChecker? Also some methods on `inclusion` struct are public which can be private as they only called by InclusionChecker methods which are exposed to other packages.

File: core/tracker/inclusion.go (hunk 3 above)
Comment (corverroos) [reply core/tracker/inclusion.go:144]: good catch, added "aggregated" logging field

File: core/tracker/inclusion.go (hunk 4 above)
Comment (corverroos) [reply core/tracker/inclusion.go:189]: Does need more stuff

File: core/tracker/inclusion.go (hunk 5 above)
Comment (corverroos) [reply core/tracker/inclusion.go:128]: it is its API

File: core/tracker/inclusion.go (hunk 6 above)
Comment (corverroos) [reply core/tracker/inclusion.go:237]: renamed it to `inclusionChecker` which wrap a `inclusionCore`. The core is decoupled and optimised for easy testing.

File: core/tracker/incldelay.go [hunk 7]
```diff
@@ -1,124 +0,0 @@
...
-				duty, ok := def.(core.AttesterDefinition)
-				if !ok {
-					return errors.New("invalid attester definition")
-				}
-
-				if duty.CommitteeIndex != att.Data.Index {
-					continue // This duty is for another committee
-				}
-
-				if !att.AggregationBits.BitAt(duty.ValidatorCommitteeIndex) {
```
Comment (dB2510) [inline core/tracker/incldelay.go:98]: @corverroos you need to add this check in new inclusion checker

Comment (corverroos) [reply core/tracker/incldelay.go:98]: done

Comment (dB2510) [reply core/tracker/incldelay.go:98]: I think you haven't pushed that commit. It might be in local commit?

File: core/tracker/inclusion.go [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+	}
+
+	log.Info(ctx, msg,
+		z.I64("block_slot", blockSlot),
+		z.I64("attestation_slot", attSlot),
+		z.Any("pubkey", sub.Pubkey),
+		z.I64("inclusion_delay", inclDelay),
+		z.Any("broadcast_delay", sub.Delay),
+		z.Int("aggregate_len", len(aggIndices)),
+		z.Bool("aggregated", len(aggIndices) > 1),
```
Comment (corverroos) [inline core/tracker/inclusion.go:229]: @dB2510 the aggregated logging field is added here

File: core/tracker/incldelay.go (hunk 7 above)
Comment (corverroos) [reply core/tracker/incldelay.go:98]: Cool, should be here now: https://github.com/ObolNetwork/charon/pull/2120/files#r1174812448

File: core/tracker/inclusion.go [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+		switch sub.Duty.Type {
+		case core.DutyAttester, core.DutyAggregator:
+			_, ok := block.Attestations[sub.AttRoot]
+			if !ok {
+				remaining = append(remaining, sub)
+				continue
+			}
+
+			// TODO(corver): We should probably check if validator included in AggregationBits
+			//  for attester duty by caching scheduler duty data to get validator committee index.
```
Comment (dB2510) [inline core/tracker/inclusion.go:158]: cool, i was looking for this. Will push in next PR probably?

File: core/tracker/inclusion.go (hunk 8 above)
Comment (dB2510) [reply core/tracker/inclusion.go:229]: Just to clarify, this aggregated logging will tell us that attestation is an aggregate attestation and NOT an unaggregated one. Well here `aggregated: true` only when we are aggregator of that slot AND it is our aggregate attestation.

File: core/tracker/inclusion.go [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+	for _, sub := range i.submissions {
+		switch sub.Duty.Type {
+		case core.DutyAttester, core.DutyAggregator:
+			_, ok := block.Attestations[sub.AttRoot]
+			if !ok {
+				remaining = append(remaining, sub)
+				continue
+			}
+
+			// TODO(corver): We should probably check if validator included in AggregationBits
```
Comment (xenowits) [inline core/tracker/inclusion.go:157]: Yeah. If we wish, we can also understand if the attestation was aggregated or not by checking
```go
if NO_OF_SET_BITS(attestation.aggregation_bits) == 1 { 
  "attestation was NOT aggregated" 
}
```

File: core/tracker/inclusion.go (hunk 9 above)
Comment (corverroos) [reply core/tracker/inclusion.go:158]: yeah

**Review by dB2510 — APPROVED** [review]

**Pull request merged**